msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInMFATest.kt

```diff
@@ -0,0 +1,180 @@
+        (sendChallengeResult as MFARequiredResult.VerificationRequired)
+        assertNotNull(sendChallengeResult.sentTo)
+        assertNotNull(sendChallengeResult.codeLength)
+        assertNotNull(sendChallengeResult.channel)
+
+        // Submit incorrect challenge
+        val submitIncorrectChallengeResult = sendChallengeResult.nextState.submitChallenge("invalid")
+        assertResult<SubmitChallengeError>(submitIncorrectChallengeResult)
+        assertTrue((submitIncorrectChallengeResult as SubmitChallengeError).isInvalidChallenge())
+
+        // Request new challenge
+        val requestNewChallengeResult = sendChallengeResult.nextState.requestChallenge()
+        assertResult<MFARequiredResult.VerificationRequired>(requestNewChallengeResult)
+        (requestNewChallengeResult as MFARequiredResult.VerificationRequired)
+        assertNotNull(requestNewChallengeResult.sentTo)
+        assertNotNull(requestNewChallengeResult.codeLength)
+        assertNotNull(requestNewChallengeResult.channel)
+
+        // Ugly fix until https://identitydivision.visualstudio.com/Engineering/_workitems/edit/3024168 is in place
+        sleep(3000)
+
+        // Retrieve challenge from mailbox and submit
+        val otp = tempEmailApi.retrieveCodeFromInbox(username)
+        val submitCorrectChallengeResult = requestNewChallengeResult.nextState.submitChallenge(otp)
+        assertResult<SignInResult.Complete>(submitCorrectChallengeResult)
+    }
+
+    /**
+     * Full flow:
+     * - Receive MFA required error from API.
+     * - Request default challenge.
+     * - Challenge sent successfully, SelectionRequired is returned.
+     * - Call getAuthMethods to retrieve all auth methods available.
+     * - Request new challenge.
+     * - Submit correct challenge.
+     * - Complete MFA flow and complete sign in.
+     */
+    @Test
+    fun `test get other auth methods, request challenge on specific auth method and complete MFA flow`() = runTest {
+        val username = config.email
+        val password = getSafePassword()
+        val result = application.signIn(username, password.toCharArray())
+        assertResult<SignInResult.MFARequired>(result)
+
+        // Initiate challenge, send code to email
+        val sendChallengeResult = (result as SignInResult.MFARequired).nextState.requestChallenge()
+        assertResult<MFARequiredResult.VerificationRequired>(sendChallengeResult)
+        (sendChallengeResult as MFARequiredResult.VerificationRequired)
+        assertNotNull(sendChallengeResult.sentTo)
+        assertNotNull(sendChallengeResult.codeLength)
+        assertNotNull(sendChallengeResult.channel)
+
+        // Retrieve other auth methods
+        val getAuthMethodsResult = sendChallengeResult.nextState.getAuthMethods()
+        assertResult<MFARequiredResult.SelectionRequired>(getAuthMethodsResult)
+        (getAuthMethodsResult as MFARequiredResult.SelectionRequired)
+        assertTrue(getAuthMethodsResult.authMethods.size == 1)
```

Comment:
On iOS we check also that the challenge type is "email"

Reply:
updated